### ISSUE TITLE
Mass bug: architecture_independent tag verification

### ISSUE THREAD

**Issue by cottsay** [ros/rosdistro#4037]
With the addition of `noarch` support for RPMs in bloom (https://github.com/ros-infrastructure/bloom/pull/270), I found a couple of packages that incorrectly marked `architecture_independent` when they were, in fact, not so. These packages were notified and the appropriate `package.xml`s updated.

Inversely, there are MANY packages that do not mark `architecture_independent` when they should. This is mostly because the current debian buildfarm is unable to leverage this.

My experimental RPM buildfarm (http://csc.mcs.sdsmt.edu/jenkins/) has the capability to both utilize this and to check when it is missing, so I'm compiling a list of packages that should be updated.

Hopefully this will act as a first step in implementing this functionality on the debian buildfarm, and to produce smaller, more accurate builds on the RPM buildfarm.

Thanks!

Originally https://github.com/ros/catkin/issues/623
## Package List (work-in-progress):

After compiling the list, I've decided that there are just too many packages to target all of Hydro. Instead, I'll try to catch Indigo packages as they come in.

This is a preliminary list. Some of these may be false positives.
- [x] `ackermann_msgs` https://github.com/jack-oquin/ackermann_msgs/pull/3
- [x] `actionlib_msgs` https://github.com/ros/common_msgs/pull/35
- [x] `angles` https://github.com/ros/angles/pull/4
- [x] `bond` https://github.com/ros/bond_core/pull/4
- [x] `bondpy` https://github.com/ros/bond_core/pull/4
- [x] `calibration_estimation` https://github.com/ros-perception/calibration/pull/24
- [x] `calibration_launch` https://github.com/ros-perception/calibration/pull/24
- [x] `calibration_msgs` https://github.com/ros-perception/calibration/pull/24
- [x] `camera_calibration` https://github.com/ros-perception/image_pipeline/commit/dac6e56c46b19474f7a18fbb40b312c6173a8ad8
- [x] `capabilities` https://github.com/osrf/capabilities/pull/76
- [x] `catkin` https://github.com/ros/catkin/pull/625
- [x] `cmake_modules` https://github.com/ros/cmake_modules/pull/18
- [x] `concert_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [x] `concert_service_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [x] `control_msgs` https://github.com/ros-controls/control_msgs/pull/11
- [x] `diagnostic_msgs` https://github.com/ros/common_msgs/pull/35
- [ ] `driver_base` https://github.com/ros-drivers/driver_common/pull/6
- [ ] `ecl_build`
- [ ] `ecl_concepts`
- [ ] `ecl_converters_lite`
- [ ] `ecl_eigen`
- [ ] `ecl_license`
- [ ] `ecl_linear_algebra`
- [ ] `ecl_math`
- [ ] `ecl_mpl`
- [ ] `eigen_stl_containers`
- [ ] `flir_ptu_description`
- [ ] `flir_ptu_viz`
- [x] `gateway_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [x] `gencpp` https://github.com/ros/gencpp/pull/19
- [ ] `geneus`
- [ ] `genlisp`
- [x] `genmsg` https://github.com/ros/genmsg/pull/71
- [x] `genpy` https://github.com/ros/genpy/pull/27
- [x] `geometry_msgs` https://github.com/ros/common_msgs/pull/35
- [ ] `household_objects_database_msgs`
- [ ] `joint_state_publisher`
- [ ] `librviz_tutorial`
- [ ] `manipulation_msgs`
- [ ] `map_msgs`
- [ ] `media_export`
- [ ] `message_generation`
- [ ] `message_runtime`
- [ ] `mk`
- [ ] `moveit_commander`
- [ ] `moveit_ikfast`
- [ ] `moveit_msgs`
- [ ] `moveit_resources`
- [x] `nav_msgs` https://github.com/ros/common_msgs/pull/35
- [ ] `nmea_msgs`
- [ ] `nodelet_topic_tools`
- [x] `object_recognition_msgs`
- [ ] `octomap_msgs`
- [ ] `pcl_conversions`
- [ ] `pcl_msgs`
- [ ] `pr2_mechanism_msgs`
- [ ] `python_qt_binding`
- [ ] `qt_dotgraph`
- [ ] `qt_gui_app`
- [ ] `qt_gui`
- [ ] `qt_gui_py_common`
- [ ] `rail_maps`
- [x] `rocon_app_manager_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [ ] `rocon_bubble_icons`
- [ ] `rocon_console`
- [x] `rocon_device_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [ ] `rocon_ebnf`
- [ ] `rocon_gateway`
- [ ] `rocon_gateway_tests`
- [ ] `rocon_gateway_utils`
- [ ] `rocon_hub_client`
- [ ] `rocon_hub`
- [ ] `rocon_icons`
- [x] `rocon_interaction_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [ ] `rocon_interactions`
- [ ] `rocon_launch`
- [ ] `rocon_master_info`
- [ ] `rocon_python_comms`
- [ ] `rocon_python_redis`
- [ ] `rocon_python_utils`
- [ ] `rocon_python_wifi`
- [ ] `rocon_semantic_version`
- [x] `rocon_service_pair_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [x] `rocon_std_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [ ] `rocon_test`
- [x] `rocon_tutorial_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [ ] `rocon_uri`
- [ ] `rosapi`
- [ ] `rosbag_migration_rule`
- [ ] `rosbash`
- [ ] `rosboost_cfg`
- [ ] `rosbridge_library`
- [ ] `rosbridge_server`
- [ ] `rosbuild`
- [ ] `rosclean`
- [ ] `roscpp_traits`
- [ ] `roscreate`
- [ ] `rosdoc_lite`
- [x] `rosgraph` https://github.com/ros/ros_comm/pull/391
- [ ] `rosgraph_msgs`
- [ ] `roslang`
- [x] `roslaunch` https://github.com/ros/ros_comm/pull/391
- [ ] `roslint`
- [ ] `roslisp`
- [ ] `rosmake`
- [x] `rosmaster` https://github.com/ros/ros_comm/pull/391
- [x] `rosmsg` https://github.com/ros/ros_comm/pull/391
- [x] `rosnode` https://github.com/ros/ros_comm/pull/391
- [x] `rosparam` https://github.com/ros/ros_comm/pull/391
- [x] `rospy` https://github.com/ros/ros_comm/pull/391
- [ ] `rospy_tutorials`
- [x] `rosservice` https://github.com/ros/ros_comm/pull/391
- [x] `rostest` https://github.com/ros/ros_comm/pull/391
- [ ] `rostopic`
- [ ] `rosunit`
- [x] `roswtf` https://github.com/ros/ros_comm/pull/391
- [x] `rqt_action` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_bag` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_bag_plugins` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_console` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_dep` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [ ] `rqt_gui`
- [ ] `rqt_gui_py`
- [x] `rqt_graph` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_launch` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_logger_level` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_msg` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_plot` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_publisher` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_py_common` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_py_console` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_reconfigure` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_service_caller` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_shell` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_srv` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_top` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_topic` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rqt_web` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `rviz_python_tutorial` https://github.com/ros-visualization/rqt_common_plugins/pull/254
- [x] `scheduler_msgs` https://github.com/robotics-in-concert/rocon_msgs/pull/95
- [x] `sensor_msgs` https://github.com/ros/common_msgs/pull/35
- [x] `shape_msgs` https://github.com/ros/common_msgs/pull/35
- [ ] `smach`
- [ ] `smach_msgs`
- [ ] `smach_ros`
- [ ] `smach_viewer`
- [x] `smclib` https://github.com/ros/bond_core/pull/4
- [ ] `std_capabilities`
- [ ] `std_msgs`
- [ ] `std_srvs`
- [x] `stereo_msgs` https://github.com/ros/common_msgs/pull/35
- [ ] `tf2_bullet`
- [ ] `tf2_geometry_msgs`
- [ ] `tf2_kdl`
- [ ] `tf2_msgs`
- [ ] `tf2_py`
- [ ] `tf2_tools`
- [ ] `timestamp_tools` https://github.com/ros-drivers/driver_common/pull/6
- [x] `trajectory_msgs` https://github.com/ros/common_msgs/pull/35
- [x] ~~`unique_id`~~ https://github.com/ros/rosdistro/issues/4037#issuecomment-41842366
- [ ] `urdfdom_headers`
- [ ] `urdfdom_py`
- [ ] `urdf_parser_plugin`
- [x] `uuid_msgs` https://github.com/ros/rosdistro/issues/4037#issuecomment-41842366
- [x] `visualization_msgs` https://github.com/ros/common_msgs/pull/35
- [x] `xacro` https://github.com/ros/xacro/pull/32
- [x] `xdot` https://github.com/jbohren/xdot/pull/4
- [x] `zeroconf_msgs` https://github.com/stonier/zeroconf_msgs/pull/1

**Comment by mikaelarguedas** [comment]
@cottsay Is this still of interest ? Is the aforementioned fedora buildfarm still active (for testing, if we keep adding the architecture independant tag to packages) ?

**Comment by cottsay** [comment]
The Fedora buildfarm is not currently active. Maybe I'll have cycles to get things running again some day, but not now.

**Comment by jack-oquin** [comment]
I recently noticed a ROS2 pull request to *remove* ``<architecture-independent/>`` from one of the unique_identifier packages. 

Is there opposite motion going on?

**Comment by mikaelarguedas** [comment]
While this tag is used only in the RPM metadata generated by bloom, I think it's valuable information and maybe the ROS buildfarm will leverage it some day. So we do not encourage users to get rid of it in ROS2.

One thing to keep in mind though is that ROS2 message packages do compile code and libraries so cannot be considered `architecture-independent` anymore. Only pure python packages (build type `ament_python` in the ROS2 world) and header-only packages can be considered architecture independent.

**Comment by tfoote** [comment]
@cottsay Maybe this would be worth returning to briefly with your current work. I audited all the ones with PRs open and was able to check most of them off as merged.

**Comment by cottsay** [comment]
I'll start by working through the list as it was 4 years ago, but I expect there have been new packages which need to be discovered. Since we don't have an RPM buildfarm active at the moment, I may have to come up with an alternate bulk discovery method.

Since the ROS2 package list is more containable right now, I might go after those to get ahead of the problem.

**Comment by tfoote** [comment]
@cottsay We've turned on RPM builds on ROS 2 rolling and we've shipped Noetic already which is the last ROS 1. To that end this seems like it should be caught now in rolling and we're not likely to turn on for Noetic so can we close out this long running ticket as enough for now?

**Comment by cottsay** [comment]
> this seems like it should be caught now in rolling

The new buildfarm code doesn't have support for RPM linting yet, so we won't catch packages that are missing the tag but should have it. We'll still catch packages that claim to be architecture-independent but actually aren't.

I have a hunch that ROS 2 is in pretty rough shape as far as this flag is concerned. There are advantages to doing `noarch` right, but we still don't have any capability in the buildfarm to build `noarch` packages only once, so we can't take full advantage of it even if it's declared correctly in the packages.

I think we should close this bug now - it was intended for tracking the feature in ROS 1, and like you said, it's unlikely that we'll ever see Noetic RPMs on the buildfarm. We can always re-open it if things change.
